cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -25,7 +25,7 @@ class Node extends Instruction {
   }
 
   /** Gets the expression corresponding to this node, if any. */
-  Expr asExpr() { result = this.getUnconvertedResultExpression() }
+  Expr asExpr() { result = this.getConvertedResultExpression() }
```

Comment:
So if we had:
```
short func(int& r) {
  return (short)r;
}
```
There are three `Expr`s of interest: the `VariableAccess` of `r`, the `ReferenceDeref`, and the `Cast` to `short`. Would there be three `ExprNode`s, with the `getExpr()` for all three returning the same `VariableAccess`? That seems confusing.

Reply:
That's exactly what I'm proposing. I think in your example it's what most users will want, keeping in mind that the data flow library is often the first thing novice users reach for, long before learning about conversions. Users that come from our other languages or from the current data flow library will expect it to "just work" when they mark the `VariableAccess` to `r` as a source.

If we turn the argument around, is there then a situation where the user would get unexpected false positives due to the behaviour I'm proposing?

I agree that there should be a predicate available that models conversions accurately, but that would be for advanced users only.

Reply:
I'm not so concerned about false positives in this case. First, I don't think it'll actually happen in practice. If a sink has type `int`, then a value of type `int&` should not be able to reach it and vice versa. If the sink itself is a lvalue-to-reference conversion, then it might happen that both the value and the reference can flow there, but what's the harm in that? Second, false negatives are much worse than false positives in security queries. Queries in our standard suite should not use the `asExpr` that I'm proposing but should use `asConvertedExpr`.

With the behaviour of `asExpr` you propose, I don't know how to write the QLDoc to recommend to users whether they should call `getFullyConverted()` or not. They should certainly do it in your `return (short)r;` example. In the case of array-to-pointer conversions they probably shouldn't if the tainted data is in the array itself. To fully understand whether to call `getFullyConverted`, they'd need to understand our `Conversion` system, and I wouldn't expect even 10% of non-Semmle QL users to understand that.

Thinking about it more, I can see it's too extreme to let `asExpr` include _any_ conversion. It's really only conversions between values and pointers/references that we want. That's array-to-pointer, value-to-reference, and reference-to-value.

Reply:
Doesn't the AST-based dataflow library create a separate `ExprNode` for each `Expr`, regardless of whether it's a conversion or not? Shouldn't we match that behavior to make it easier to port existing dataflow rules?

Reply:
I agree we should match the existing behaviour. As I remember it, it’s completely unaware of conversions. That means there is a Node for each Conversion, but such nodes have no flow from them or to them.

Reply:
When I say "match the existing behaviour" I really mean allow the same kind of caller to work the same way.

If it seems too confusing to have multiple instructions map to the same `Expr`, we can also prescribe one of those instructions to be the right choice. Unless you can think of reasonable use cases for tracking the reference, rather than its contents, we can make it the default choice to track contents. I think it's easier to state what I mean in terms of `DataFlow::exprNode`, which would be a total and injective function from non-conversions to `Node`.

1. If a non-conversion `e` has a `ReferenceDereferenceExpr c` directly on it, then `exprNode(e)` is the instruction of `c`. As far as I can tell, `ReferenceDereferenceExpr` always comes first in the chain of conversions.
2. If a non-conversion `e` has an `ArrayToPointerConversion c` in its conversion chain, then `exprNode(e)` is the instruction of `c.getExpr()`.
3. If a non-conversion `e` has a `ReferenceToExpr c` in its conversion chain, then `exprNode(e)` is the instruction of `c.getExpr()`.

Reply:
I think I agree with your suggestion. However, as I started to implement the predicate to select the "appropriate" `Expr`, I hit enough corner cases that I'm not comfortable merging that change for 1.19. For example:
```
int a[10];
int* p = (int (&)[10])(float (&)[10])a;
``` 
I believe that expression has both an `ArrayToPointerConversion` and a `ReferenceDereferenceExpr`, with the `ReferenceDereferenceExpr` not being the innermost `Conversion`.

For now, I've simply stopped creating an `ExprNode` for a `Conversion` at all, which should be close to the AST-based behavior. I've opened CPP-318 to do the right thing.

Reply:
What a terrifying edge case you have there.

I don't see how it helps to stop creating `ExprNode` for `Conversion`s. Callers that use this library like the current data flow library, never calling `getFullyConverted`, will get the same result as before. Callers that know about conversions will get worse results.

Reply:
OK, I've reverted that commit.